### ISSUE TITLE
[$250][HOLD #49825][HOLD #49824] Chat - Chat is not autoscrolled to the bottom when receiving whisper after a mention

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#49599]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.39.0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Open the staging.new.expensify.com website
2. Open any room chat in inbox
3. Mention another user that is not a member of that room
4. Verify that a whisper is received afrer sending the mention, and the chat is autoscrolled to the bottom

## Expected Result:
After mentioning another user that is not part of the chat, a whisper with options should be received, and the chat should be autoscrolled to the bottom

## Actual Result:
When the whisper is received, after mentioning an user that is not part of the room, the chat is not autoscrolled to the bottom

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/user-attachments/assets/931f56f7-bbad-42ba-9a1a-705d221436ff

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021838383119719075523</li>
        <li>Upwork Job ID: 1838383119719075523</li>
        <li>Last Price Increase: 2024-09-24</li>
<li>Automatic offers: </li>
<ul>
<li>paultsimura | Reviewer | 104109892</li>
<li>klajdipaja | Contributor | 104109896</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @dangrous</details>

**Comment by lanitochka17** [comment]
@RachCHopkins  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by RachCHopkins** [comment]
Fine on web/desktop. 

Same on iOS native app. Not just android/web, my guess is that it's anything with a mobile aspect ratio.

![image](https://github.com/user-attachments/assets/7855898b-f47a-4b0c-af56-a47668db9b85)

**Comment by klajdipaja** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When a whisper is received in the chat for the current user the screen is not scrolled down and the whisper is not visible until the user decides to scroll down.

### What is the root cause of that problem?
The cause of the issue seems to be an intented behaviour in the logic of the scrolling for new actions. The current logic scrolls the user to the bottom if the action is from the current user, all other cases no scrolling is done. This logic is in `ReportActionsList.tsx` in this code block:

```
    const scrollToBottomForCurrentUserAction = useCallback(
        (isFromCurrentUser: boolean) => {
            // If a new comment is added and it's from the current user scroll to the bottom otherwise leave the user positioned where
            // they are now in the list.
            if (!isFromCurrentUser) {
                return;
            }
            if (!hasNewestReportActionRef.current) {
                Navigation.navigate(ROUTES.REPORT_WITH_ID.getRoute(report.reportID));
                return;
            }
            InteractionManager.runAfterInteractions(() => reportScrollManager.scrollToBottom());
        },
        [reportScrollManager, report.reportID],
    );
 ```


### What changes do you think we should make in order to solve the problem?
We have a few options depending on how the scrolling is intented to behave;
1. Always scroll down for all new actions not just for current user messages. This is a very easy fix, we just remove the `!isFromCurrentUser` check.
2. Scroll down only for messages from the current user and whispers of the current user. First part is the existing logic, we need to add another code block that would check if the action is a whisper for the current user.
After the `isFromCurrentUser` check would be replaced by the following block:

```
 const newestAction = sortedVisibleReportActions[0];
            const whisperedTo = ReportActionsUtils.getWhisperedTo(newestAction);

            const transactionsWithReceipts = ReportUtils.getTransactionsWithReceipts(report.reportID);
            const isWhisper = whisperedTo.length > 0 && transactionsWithReceipts.length === 0;
            const isWhisperOnlyVisibleByUser = isWhisper && ReportUtils.isCurrentUserTheOnlyParticipant(whisperedTo);
          
             const shouldScrollDown = isFromCurrentUser || isWhisperOnlyVisibleByUser;
            if (!shouldScrollDown) {
                return;
            }

           
```
### What alternative solutions did you explore? (Optional)

**Comment by blimpich** [comment]
Agreed, lets got with solution 2 outlined in [the proposal](https://github.com/Expensify/App/issues/49599#issuecomment-2370882042).

If possible, can we reach out the engineers who implemented this code originally to make sure they're aware of the changes we're making?

**Comment by klajdipaja** [comment]
@paultsimura While testing my changes in android emulator I found out that there's an issue with my proposal. When the internet or device is really slow I saw that the scrolling does not happen. 
The root cause of that is that the `scrollToBottomForCurrentUserAction` is called as a callback after the API.write for the action that the user has executed but if the internet is slow the whisper message has not arrived yet. 
We need to change the approach to check the latest action in the chat and check if that is a whisper for the current user or not by changing [this  existing useEffect](https://github.com/Expensify/App/blob/b1720c7a420a5f5e5482f40b8a6a31459c909d18/src/pages/home/report/ReportActionsList.tsx#L291) to this:

```
    useEffect(() => {
        const isWhisperForUser = isWhisperVisibleOnlyToUser(sortedVisibleReportActions[0])
        const shouldScrollForWhisper = isWhisperForUser && hasNewestReportAction;

        if (
            (scrollingVerticalOffset.current < AUTOSCROLL_TO_TOP_THRESHOLD &&
            previousLastIndex.current !== lastActionIndex &&
            reportActionSize.current > sortedVisibleReportActions.length &&
            hasNewestReportAction) || shouldScrollForWhisper
        ) {
            reportScrollManager.scrollToBottom();
        }
        previousLastIndex.current = lastActionIndex;
        reportActionSize.current = sortedVisibleReportActions.length;
    }, [lastActionIndex, sortedVisibleReportActions, reportScrollManager, hasNewestReportAction, linkedReportActionID]);

````


This is testing really well, you can check this out in [this branch](https://github.com/klajdipaja/Expensify-App/tree/bugfix/GH-49599_Option2_Chat_is_not_autoscrolled_to_the_bottom_when_receiving_whisper_after_a_mention) and let me know what you think

**Comment by klajdipaja** [comment]
@paultsimura did you check those with the draft PR I created? I have a feeling it will solve both issues but haven't had the time to double check

**Comment by klajdipaja** [comment]
@paultsimura Just to give you an update. 
I did a lot of debugging with this and found out that there's an issue with the `hasNewestReportAction` that I think it's rooted in the optimistic data but so far have not been able to exactly confirm this as that part of the codebase is masive.
I logged in the console the `report.lastVisibleActionCreated` and I noticed that it is sometimes off by a few nanoseconds after the comment has been submitted and later when the opimistic data is updated it goes to the correct value. This means that the `action.created` in this cases is bigger then the `report.lastVisibleActionCreated` which is the change that I added into my draft branch and now it seems to be working as expected.

I replaced  `hasNewestReportAction = sortedVisibleReportActions[0]?.created === report.lastVisibleActionCreated`  with: 
`hasNewestReportAction = sortedVisibleReportActions[0]?.created >= report.lastVisibleActionCreated`

**Comment by klajdipaja** [comment]
Updated proposal: https://github.com/Expensify/App/issues/49599#issuecomment-2370882042 
Now I have included option 3 that is implemented in the Draft branch and covers the issue in this ticket

**Comment by klajdipaja** [comment]
Regarding the slight difference of the lastVisibleActionCreated time can we have a BE engineer feedback how they set that time on their side? 
I found out the different value comes from the BE and it's also different to the clientCreatedTime which I was assuming would be what they use on their end to set the last:

I'm attaching request and response screenshots so that you can see the difference in the values.
Request:
![image](https://github.com/user-attachments/assets/5ac758cd-847d-4841-8747-841d0ae328a1)
Response:
![image](https://github.com/user-attachments/assets/2dbf28ec-a0bc-4ceb-b6a6-2ae83bb36a45)

**Comment by blimpich** [comment]
Looking at the backend we mainly use `clientCreatedTime` to determine whether or not the user has "read" the comment. I think `lastVisibleActionCreated` corresponds to when the report action was created in the DB.

**Comment by RachCHopkins** [comment]
Changing to Weekly, since we're waiting on other issues.

**Comment by RachCHopkins** [comment]
Looks like we have PRs on the boil for both linked issues.

**Comment by dangrous** [comment]
Will likely be able to be joint fixed with https://github.com/Expensify/App/issues/50803

**Comment by dangrous** [comment]
Will do! Still confirming expected behavior - whispers are weird, haha

**Comment by dangrous** [comment]
I have a potential fix up, will fully test tomorrow and then hopefully have it up for review! It'll actually be a separate fix from that other issue, but that's okay.

**Comment by RachCHopkins** [comment]
How this going @dangrous?

**Comment by dangrous** [comment]
Backend is in review! Should hopefully be merged soon and then we can move forward here

**Comment by dangrous** [comment]
This should hopefully be fixed on the backend. Can we retest and figure out if anything else is needed?

**Comment by dangrous** [comment]
so for [the first one](https://github.com/Expensify/App/issues/49599#issuecomment-2477468658) I think that might be expected? What's the behavior if someone sends a regular message when you're scrolled up in the chat, like you are? I don't think you should be rescrolled to the bottom, right?

For the second one - is that new behavior, or was it like that before too? It definitely seems weird to have them all pop back up like that - but I'm not sure why they disappeared in the first place. This would probably be an edge case since you'd be unlikely to keep inviting the same person over and over again like that, but still worth looking into. Is it the front end that's removing the previous whispers (maybe optimistically?) or the backend?

**Comment by dangrous** [comment]
Okay great, then I think we're good to close this! Since no PR raised, I don't think we need payment here but if anyone disagrees let me know. Thanks!

**Comment by klajdipaja** [comment]
@dangrous @paultsimura what should be done with the Upwork contract in this case?

**Comment by dangrous** [comment]
actually, I've been informed that we pay 50% in this case (contributor hired but no PR needed)! @RachCHopkins can you help us out?

**Comment by RachCHopkins** [comment]
Payment Summary:

- Contributor:  @klajdipaja to be paid $125 via Upwork
- Contributor+: @paultsimura to paid $125 via Upwork

Upwork job [here](https://www.upwork.com/jobs/~021838383119719075523)

**Comment by dangrous** [comment]
I think we can pay whenever, since it was only backend we're not waiting on a regression period cc @RachCHopkins

**Comment by RachCHopkins** [comment]
@paultsimura it's ASAP! I'm just getting my ducks in a row! 🦆 🦆🦆

**Comment by RachCHopkins** [comment]
Contributors have been paid, the contracts have been completed, and the Upwork post has been closed.
